### PR TITLE
REST: add support for multipart

### PR DESCRIPTION
**Pull request by stephan-merkli**
323842

### CODE REVIEW

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartMessage.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ *     .accept(MediaType.APPLICATION_JSON)
+ *     .post(multiPartMessage.toEntity(), TestDo.class);
+ * </pre>
+ */
+@Bean
+public class MultipartMessage {
+
+  public static final String BOUNDARY_PARAMETER = "boundary";
+
+  private List<MultipartPart> m_parts = new ArrayList<>();
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartMessage.java]: I would changing the visibility to protected

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartMessage.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    if (m_boundary == null) {
+      // create a new unique boundary if it wasn't already cached before for this multipart message
+      m_boundary = BEANS.get(IUuidProvider.class).createUuid().toString().replace("-", "");
+    }
+    return new MediaType(MediaType.MULTIPART_FORM_DATA_TYPE.getType(), MediaType.MULTIPART_FORM_DATA_TYPE.getSubtype(), CollectionUtility.hashMap(ImmutablePair.of(BOUNDARY_PARAMETER, m_boundary)));
+  }
+
+  @Override
+  public String toString() {
+    return new StringJoiner(", ", MultipartMessage.class.getSimpleName() + "[", "]")
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartMessage.java]: Usually in Scout toString() is written like this:

```
  @Override
  public String toString() {
    return DataObjectAttributeDescriptor.class.getSimpleName() + " [name=" + m_name + " type=" + m_type + " formatPattern=" + m_formatPattern + " accessor=" + m_accessor + "]";
  }
```

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartMessage.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+import org.eclipse.scout.rt.platform.BEANS;
+import org.eclipse.scout.rt.platform.Bean;
+import org.eclipse.scout.rt.platform.util.CollectionUtility;
+import org.eclipse.scout.rt.platform.util.ImmutablePair;
+import org.eclipse.scout.rt.platform.util.uuid.IUuidProvider;
+import org.eclipse.scout.rt.rest.client.IRestClientHelper;
+
+/**
+ * To be used on the sending side of a multipart request, e.g. by using a {@link IRestClientHelper}.
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartMessage.java]: Suggestion:

"Multipart message for REST service invocations using a multi part request..."

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartPart.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  public String getContentType() {
+    return m_contentType;
+  }
+
+  public InputStream getInputStream() {
+    return m_inputStream;
+  }
+
+  @Override
+  public String toString() {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartPart.java:110]: See comment on MultipartMessage.toString()

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/IMultipartMessage.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ */
+package org.eclipse.scout.rt.rest.multipart;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+
+import org.eclipse.scout.rt.rest.IRestResource;
+
+/**
+ * To be used on the receiving side of a multipart request, i.e. within a {@link IRestResource}.
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/IMultipartMessage.java]: Suggestion:

"Multipart message for receiving a REST service invocation using a multi part request..."

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/IMultipartPart.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.io.InputStream;
+import java.util.Iterator;
+
+import org.eclipse.scout.rt.platform.util.IOUtility;
+
+/**
+ * A part as returned by the {@link Iterator} of {@link IMultipartMessage}.
+ * <p>
+ * try-finally is required, {@link AutoCloseable} is used to close the {@link #getInputStream()}.
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/IMultipartPart.java]: Add  "**Note:** try-finally [...]"

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/MultipartMessageBodyReader.java [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyReader;
+import javax.ws.rs.ext.MessageBodyWriter;
+
+import org.eclipse.scout.rt.platform.exception.PlatformException;
+import org.eclipse.scout.rt.platform.util.StringUtility;
+import org.eclipse.scout.rt.rest.RestApplication.IRestApplicationClassesContributor;
+
+/**
+ * {@link MessageBodyWriter} for {@link IMultipartMessage} used on the server side (receiving).
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/MultipartMessageBodyReader.java]: Wrong referent to Writer...

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/MultipartMessageBodyReader.java [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+    /**
+     * {@link PartInputStream} based on {@link #m_inputStream}.
+     */
+    private PartInputStream m_partInputStream;
+
+    public ServerMultipartMessage(MediaType mediaType, InputStream inputStream) {
+      // boundary used within message is prefixed by -- (see https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html)
+      String boundary = "--" + assertNotNull(mediaType.getParameters().get("boundary"), "boundary parameter in media type is required");
+      // boundary must always be in 7-bit ASCII (https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html), still use UTF-8, same bytes then
+      m_newlineBoundaryBytes = ("\r\n" + boundary).getBytes(StandardCharsets.UTF_8);
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/MultipartMessageBodyReader.java:106]: is "\r\n" platform-independant or should a "\n" without "\r" also be considered?

File: org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/multipart/MultipartMessageBodyWriter.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+ * module is used instead to hide implementation details.
+ * <p>
+ * Idea from <a href="https://guntherrotsch.github.io/blog_2021/jaxrs-multipart-client.html">JAX/RS Multipart Client by
+ * Gunther Rotsch</a> released under Apache License 2.0. Scout uses a different approach via direct stream processing
+ * instead of working with temporary files. {@link StandardCharsets#UTF_8} encoding is used instead of
+ * {@link StandardCharsets#US_ASCII}.
+ */
+public class MultipartMessageBodyWriter implements MessageBodyWriter<MultipartMessage> {
+
+  private static final String HTTP_LINE_DELIMITER = "\r\n";
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/multipart/MultipartMessageBodyWriter.java:54]: platform-independent?

**Review by paolobazzi — APPROVED** [review]
Looks good! Some minor suggestions in comments

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartPart.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+   *
+   * @param partName
+   *          Name of part (mandatory)
+   * @param filename
+   *          Filename (recommended because content type is deduced from filename extension, otherwise
+   *          {@link MediaType#APPLICATION_OCTET_STREAM} is used)
+   * @param inputStream
+   *          Input stream (mandatory)
+   */
+  public static MultipartPart ofFile(String partName, String filename, InputStream inputStream) {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/client/multipart/MultipartPart.java:75]: Suggestion: add unit test for ofFile() using null as filename -> should result in content type "APPLICATION_OCTET_STREAM"

File: org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/MultipartMessageBodyReader.java (hunk 8 above)
Comment (stephan-merkli) [reply org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/multipart/MultipartMessageBodyReader.java:106]: Multipart uses always \r\n, see https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html

File: org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/multipart/MultipartMessageBodyWriter.java (hunk 9 above)
Comment (stephan-merkli) [reply org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/multipart/MultipartMessageBodyWriter.java:54]: Multipart uses always \r\n, see https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html
